### PR TITLE
Make witness v0 outputs non-standard

### PR DESCRIPTION
**Pull request by jl2012**
Witness v0 outputs are unsafe before segwit is activated and should remain non-standard

### CODE REVIEW

**Comment by petertodd** [comment]
Did you try spending one? They can't be spent due to the CLEANSTACK rule.

**Comment by jl2012** [comment]
@petertodd : yes, when spending it is non-standard for <= 0.12. The issue here is transactions with such _outputs_ are relayed and mined by 0.13, even before activation of segwit

**Comment by instagibbs** [comment]
concept ACK

**Comment by petertodd** [comment]
@jl2012 Ah sorry, misread that. Still there's no harm to those outputs being created.

**Comment by instagibbs** [comment]
@petertodd miner confiscation?

**Comment by petertodd** [comment]
@instagibbs There's lots of ways you can screw up and lose your coins you know... :)

**Comment by jl2012** [comment]
@petertodd why do we have IsStandard check for outputs at all? To prevent UTXO spam or to prevent people losing money by doing stupid things? This patch is for the latter reason.

**Comment by btcdrak** [comment]
Concept ACK

**Comment by petertodd** [comment]
eh, whatever, harmless to do this too.

Concept ACK

**Comment by sipa** [comment]
utACK, though we should have an issue to revisit this for 0.13.1

**Comment by NicolasDorier** [comment]
This will not be in 0.13, and 0.13.1 will include segwit, and we don't want to require people to update to 0.13.2 for starting to relay segwit transaction, so I don't understand the reason of this PR.

**Comment by NicolasDorier** [comment]
And if it is meant for 0.13, then it means that once segwit is activated, the relay of segwit outputs will be penalized because of old 0.13 nodes.

**Comment by luke-jr** [comment]
Should we make this conditional on !testnet?

@NicolasDorier This makes 0.13 equivalent to 0.12 in that regard.

**Comment by sipa** [comment]
I'd prefer to make this based on IsWitnessEnabled, so we don't need to touch that code again.

**Comment by NicolasDorier** [comment]
@luke-jr yes, but I feel it is kind of a waste, would be nice that when segwit get activated that it can be broadly relayed. Agree on using IsWitnessEnabled.

**Comment by instagibbs** [comment]
ok, concept NACK this in favor of IsWitnessEnabled

**Comment by jl2012** [comment]
updated to using IsWitnessEnabled

**Comment by sdaftuar** [comment]
Concept ACK, will test.

File: src/main.cpp [hunk 1]
```diff
@@ -1144,13 +1144,14 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
     }
 
     // Reject transactions with witness before segregated witness activates (override with -prematurewitness)
-    if (!GetBoolArg("-prematurewitness",false) && !tx.wit.IsNull() && !IsWitnessEnabled(chainActive.Tip(), Params().GetConsensus())) {
+    bool witnessenabled = IsWitnessEnabled(chainActive.Tip(), Params().GetConsensus());
```
Comment (dcousens) [inline src/main.cpp]: `s/witnessenabled/witnessEnabled` would be more consistent with all other naming, no?

Comment (jl2012) [inline src/main.cpp]: fixed

**Comment by dcousens** [comment]
utACK 1ffaff2

**Comment by NicolasDorier** [comment]
utACK 1ffaff2f747af683513d6d74a7241d41e3f6e051

**Comment by sdaftuar** [comment]
Tested ACK 1ffaff2f747af683513d6d74a7241d41e3f6e051

FYI I added a test to `p2p-segwit.py` to cover this case, feel free to include if you like:
https://github.com/sdaftuar/bitcoin/commit/21c59eac187af8a35abc9aec87bfbf51d812a449

**Comment by jl2012** [comment]
@sdaftuar I have added your test. Thank you!

**Comment by laanwj** [comment]
utACK c59c434, thanks @sdaftuar for adding the test

**Comment by laanwj** [comment]
Backported to 0.13 as f84ee3d 4f7f531 (used two commits instead of squashing because of separate authors)

**Comment by MarcoFalke** [comment]
@laanwj Removed the label as well.

**Comment by dcousens** [comment]
Heh,  title should probably read "Make witness v0 outputs non-standard until activation".

**Pull request merged**